lms/services/canvas.py

```diff
@@ -1,41 +1,198 @@
+
+        except CanvasAPIPermissionError:
+            # If we've already mapped this once, we aren't going to get a
+            # better answer now. So bail out
+            if already_mapped:
+                raise
+
+            if not (
+                mapped_id := self._match_file_in_course(
+                    file_id, course_id, file_mapping
+                )
+            ):
+                # We don't know why this failed, could be a course copy, could
+                # just be a permissions issue, all we know is we can't fix it
+                raise
+
+        return self.api.public_url(mapped_id)
+
+    @lru_cache
+    def _file_mapping(self, resource_link_id):
+        assignment = self._assignment_service.get(
+            tool_consumer_instance_guid=self._application_instance.tool_consumer_instance_guid,
+            resource_link_id=resource_link_id,
+        )
+        return CanvasFileMapping(assignment.extra)
+
+    @cached_property
+    def _application_instance(self):
+        return self._application_instance_service.get()
+
+    @lru_cache
+    def _files_in_course(self, course_id):
+        return self.api.list_files(course_id)
+
+    def _is_file_in_course(self, file_id, course_id):
+        LOG.debug("Checking if Canvas file %s is in course %s", file_id, course_id)
 
-    def _file_in_course(self, file_id, course_id):
         return any(
-            str(file_["id"]) == str(file_id) for file_ in self.api.list_files(course_id)
+            str(file_["id"]) == str(file_id)
+            for file_ in self._files_in_course(course_id)
         )
 
+    def _match_file_in_course(self, file_id, course_id, file_mapping):
+        """
+        Find matches in the current course for the specified files.
+
+        And update the mapping to match
+
+        :param file_id: The original file to match
+        :param course_id: The course to find a matching file in
+        :return: A mapped file, or None
+        """
+
+        file_ids_to_match = [file_id]
+
+        # We match not only the original id, but also the mapped id too. This
+        # allows us to code with some low probability scenarios where we have
+        # a mapping, the source and target of the mapping are inaccessible,
+        # and the source's name or size has changed. In this case we can use
+        # the information from the target to try and find a new matching file.
+        if target := file_mapping.get(file_id):
+            file_ids_to_match.append(target)
+
+        files_to_match = (
+            self._db_session.query(File)
+            .filter(
+                and_(
+                    File.application_instance_id == self._application_instance.id,
+                    File.lms_id.in_(file_ids_to_match),
+                    File.type == "canvas_file",
+                )
+            )
+            .all()
+        )
+
+        if files_to_match:
+            matching_keys = set((file.name, file.size) for file in files_to_match)
```

Comment:
In 99.9% of cases these are going to be the same, but it's conceivable they could be different if one or the other file was modified in some way. Can you even rename things in Canvas?